src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs

```diff
@@ -181,35 +221,103 @@ public static ReflectionAIFunction Build(MethodInfo method, object? target, AIFu
             return new(functionDescriptor, target, options);
         }
 
+        public static ReflectionAIFunction Build(
+            MethodInfo method,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type targetType,
+            AIFunctionFactoryOptions options)
+        {
+            _ = Throw.IfNull(method);
+
+            if (method.ContainsGenericParameters)
+            {
+                Throw.ArgumentException(nameof(method), "Open generic methods are not supported");
+            }
+
+            if (method.IsStatic)
+            {
+                Throw.ArgumentException(nameof(method), "The method must be an instance method.");
+            }
+
+            if (method.DeclaringType is { } declaringType &&
+                !declaringType.IsAssignableFrom(targetType))
+            {
+                Throw.ArgumentException(nameof(targetType), "The target type must be assignable to the method's declaring type.");
+            }
+
+            return new(ReflectionAIFunctionDescriptor.GetOrCreate(method, options), targetType, options);
+        }
+
         private ReflectionAIFunction(ReflectionAIFunctionDescriptor functionDescriptor, object? target, AIFunctionFactoryOptions options)
         {
             FunctionDescriptor = functionDescriptor;
             Target = target;
             AdditionalProperties = options.AdditionalProperties ?? EmptyReadOnlyDictionary<string, object?>.Instance;
         }
 
+        private ReflectionAIFunction(
+            ReflectionAIFunctionDescriptor functionDescriptor,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type targetType,
+            AIFunctionFactoryOptions options)
+        {
+            FunctionDescriptor = functionDescriptor;
+            TargetType = targetType;
+            AdditionalProperties = options.AdditionalProperties ?? EmptyReadOnlyDictionary<string, object?>.Instance;
+        }
+
         public ReflectionAIFunctionDescriptor FunctionDescriptor { get; }
         public object? Target { get; }
+        public Type? TargetType { get; }
```

Comment:
Why doesn't this have 

```suggestion
        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]
        public Type? TargetType { get; }
```

Reply:
I was just answering the previous question about this - I think we need the annotation here:
The trimmer analysis doesn't do "Cross method" analysis - it requires that the annotations are on all "input/output" to a single method. In this case the knowledge that something can be constructed through reflection come into the `Build` method, then it's transfered to the parameter of the ReflectionAIFunction method and then it's tranfered to the TargetType property which is then used by the InvokeCoreAsync method which is the place which actually uses the fact that the type can be constructed through reflection (by calling Activator.CreateInstance on it).

Because the analysis is per single method, the annotation needs to be on all the things which transfer the type value from the input (where we declare that it requires this) to the output (the code which uses the fact).

Reply:
It did have it. @eiriktsarpalis asked me to remove it here:
https://github.com/dotnet/extensions/pull/6193#pullrequestreview-2716543497

Reply:
> Because the analysis is per single method, the annotation needs to be on all the things which transfer the type value from the input (where we declare that it requires this) to the output (the code which uses the fact).

Does that mean there's a bug in the analyzer, as this isn't triggering a warning?

Reply:
Looks like the AOT test caught this:

https://dev.azure.com/dnceng-public/public/_build/results?buildId=994461&view=logs&j=9db12f06-3bef-52cb-1e93-45aadd48533c&t=04f10100-7f53-527f-69e4-9793d0a279c7&l=90

```
  Generating native code
/mnt/vss/_work/1/s/src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs(290): Trim analysis error IL2072: Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.<InvokeCoreAsync>d__26.MoveNext(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in call to 'System.Activator.CreateInstance(Type)'. The return value of method 'Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.TargetType.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/mnt/vss/_work/1/s/test/Libraries/Microsoft.Extensions.AotCompatibility.TestApp/Microsoft.Extensions.AotCompatibility.TestApp.csproj]
/mnt/vss/_work/1/s/src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs(290): Trim analysis error IL2072: Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.<InvokeCoreAsync>d__26.MoveNext(): 'instanceType' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors' in call to 'Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(IServiceProvider,Type,Object[])'. The return value of method 'Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.TargetType.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/mnt/vss/_work/1/s/test/Libraries/Microsoft.Extensions.AotCompatibility.TestApp/Microsoft.Extensions.AotCompatibility.TestApp.csproj]
```

Reply:
> Can you try to run it through a trimmer

We do that in CI, and it just failed.

```
/mnt/vss/_work/1/s/src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs(290): Trim analysis error IL2072: Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.<InvokeCoreAsync>d__26.MoveNext(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in call to 'System.Activator.CreateInstance(Type)'. The return value of method 'Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.TargetType.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/mnt/vss/_work/1/s/test/Libraries/Microsoft.Extensions.AotCompatibility.TestApp/Microsoft.Extensions.AotCompatibility.TestApp.csproj]
/mnt/vss/_work/1/s/src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs(290): Trim analysis error IL2072: Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.<InvokeCoreAsync>d__26.MoveNext(): 'instanceType' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors' in call to 'Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(IServiceProvider,Type,Object[])'. The return value of method 'Microsoft.Extensions.AI.AIFunctionFactory.ReflectionAIFunction.TargetType.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/mnt/vss/_work/1/s/test/Libraries/Microsoft.Extensions.AotCompatibility.TestApp/Microsoft.Extensions.AotCompatibility.TestApp.csproj]
```

I'll revert the commit that removed the attribute.

We should also fix the analyzer :)

Reply:
> why doesn't this produce any warnings during publish?

`public Repro? Create()` is getting trimmed becuase it is not getting called.

Reply:
Because the `Repro.Create` method is removed - nothing calls it. If you modify it like:
```csharp
Console.WriteLine(Create(typeof(HttpClient)).Create());
```

Then you get (publish with PublishTrimmed=true):
```
    F:\repro\illink\t2\Program.cs(20,13): Trim analysis warning IL2072: Repro.Create(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in call to 'System.Activator.CreateInstance(Type)'. The return value of method 'Repro.TargetType.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

But the analyzer still doesn't report anything, which I think it definitely should in this case.

@sbomer - this is the problem I sent you an email about - or at least one version of it. My guess is that the analyzer doesn't understand pattern matching at all, and gives up on it. Here it's really only used to basically do a null check, but it contains a "hidden" value assignment which we probably don't track.

Reply:
Face slap. I neglected to call Create after calling the ctor. Thanks.

Reply:
https://github.com/dotnet/runtime/issues/113922